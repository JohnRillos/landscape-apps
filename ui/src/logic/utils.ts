import ob from 'urbit-ob';
import { unixToDa } from '@urbit/api';
import { formatUv } from '@urbit/aura';
import anyAscii from 'any-ascii';
import { format, differenceInDays } from 'date-fns';
import _ from 'lodash';
import { ChatWhom } from '../types/chat';
import {
  Cabal,
  Cabals,
  Cordon,
  Group,
  PrivacyType,
  Rank,
} from '../types/groups';

export function renderRank(rank: Rank, plural = false) {
  if (rank === 'czar') {
    return plural ? 'Galaxies' : 'Galaxy';
  }
  if (rank === 'king') {
    return plural ? 'Stars' : 'Star';
  }
  if (rank === 'duke') {
    return plural ? 'Planets' : 'Planet';
  }
  if (rank === 'earl') {
    return plural ? 'Moons' : 'Moon';
  }
  return plural ? 'Comets' : 'Comet';
}

/**
 * Processes a string to make it `@tas` compatible
 */
export function strToSym(str: string): string {
  const ascii = anyAscii(str);
  return ascii.toLowerCase().replaceAll(/[^a-zA-Z0-9-]/g, '-');
}

export function channelHref(flag: string, ch: string) {
  return `/groups/${flag}/channels/chat/${ch}`;
}

export function makePrettyDay(date: Date) {
  const diff = differenceInDays(new Date(), date);
  switch (diff) {
    case 0:
      return 'Today';
    case 1:
      return 'Yesterday';
    default:
      return `${format(date, 'LLLL')} ${format(date, 'do')}`;
  }
}

export function makePrettyDayAndTime(date: Date) {
  const diff = differenceInDays(new Date(), date);
  const time = format(date, 'HH:mm');
  switch (true) {
    case diff === 0:
      return `Today • ${time}`;
    case diff === 1:
      return `Yesterday • ${time}`;
    case diff > 1 && diff < 8:
      return `${format(date, 'cccc')} • ${time}`;
    default:
      return `${format(date, 'LLLL')} ${format(date, 'do')} • ${time}`;
  }
}

export function makePrettyDayAndDateAndTime(date: Date) {
  const diff = differenceInDays(new Date(), date);
  const time = format(date, 'HH:mm');
  const fullDate = `${format(date, 'LLLL')} ${format(date, 'do')}, ${format(
    date,
    'yyyy'
  )}`;
  switch (true) {
    case diff === 0:
      return `Today • ${time} • ${fullDate}`;
    case diff === 1:
      return `Yesterday • ${time} • ${fullDate}`;
    case diff > 1 && diff < 8:
      return `${format(date, 'cccc')} • ${time} • ${fullDate}`;
    default:
      return `${fullDate} • ${time}`;
  }
}

export function whomIsDm(whom: ChatWhom): boolean {
  return whom.startsWith('~') && !whom.match('/');
}

// ship + term, term being a @tas: lower-case letters, numbers, and hyphens
export function whomIsFlag(whom: ChatWhom): boolean {
  return (
    /^~[a-z-]+\/[a-z]+[a-z0-9-]*$/.test(whom) &&
    ob.isValidPatp(whom.split('/')[0])
  );
}

export function whomIsMultiDm(whom: ChatWhom): boolean {
  return whom.startsWith(`0v`);
}

export function normalizeUrbitColor(color: string): string {
  if (color.startsWith('#')) {
    return color;
  }

  const colorString = color.slice(2).replace('.', '').toUpperCase();
  const lengthAdjustedColor = _.padEnd(colorString, 6, _.last(colorString));
  return `#${lengthAdjustedColor}`;
}

export function pluralize(word: string, count: number): string {
  if (count === 1) {
    return word;
  }

  return `${word}s`;
}

export function createStorageKey(name: string): string {
  return `~${window.ship}/${window.desk}/${name}`;
}

// for purging storage with version updates
export function clearStorageMigration<T>() {
  return {} as T;
}

export const storageVersion = parseInt(
  import.meta.env.VITE_STORAGE_VERSION,
  10
);

export function preSig(ship: string): string {
  if (!ship) {
    return '';
  }

  if (ship.trim().startsWith('~')) {
    return ship.trim();
  }

  return '~'.concat(ship.trim());
}

export function newUv(seed = Date.now()) {
  return formatUv(unixToDa(seed));
}

export function getSectTitle(cabals: Cabals, sect: string) {
  return cabals[sect]?.meta.title || sect;
}

export function getFlagParts(flag: string) {
  const parts = flag.split('/');

  return {
    ship: parts[0],
    name: parts[1],
  };
}

export function isValidUrl(str?: string): boolean {
  const pattern = new RegExp(
    '^(https?:\\/\\/)?' + // protocol
      '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
      '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
      '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
      '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
      '(\\#[-a-z\\d_]*)?$',
    'i'
  ); // fragment locator
  return str ? !!pattern.test(str) : false;
}

export function getGroupPrivacy(cordon: Cordon): PrivacyType {
  if ('open' in cordon) {
    return 'public';
  }

  if ('shut' in cordon) {
    return 'private';
  }

  return 'secret';
}

export function toTitleCase(s: string): string {
  if (!s) {
    return '';
  }
  return s
    .split(' ')
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ');
}

export function randomElement<T>(a: T[]) {
  return a[Math.floor(Math.random() * a.length)];
}
