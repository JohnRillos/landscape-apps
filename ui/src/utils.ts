import anyAscii from 'any-ascii';
import { format, differenceInDays } from 'date-fns';
import { ChatWhom } from './types/chat';
import { Rank } from './types/groups';

export function renderRank(rank: Rank, plural = false) {
  if (rank === 'czar') {
    return plural ? 'Galaxies' : 'Galaxy';
  }
  if (rank === 'king') {
    return plural ? 'Stars' : 'Star';
  }
  if (rank === 'duke') {
    return plural ? 'Planets' : 'Planet';
  }
  if (rank === 'earl') {
    return plural ? 'Moons' : 'Moon';
  }
  return plural ? 'Comets' : 'Comet';
}

/**
 * Processes a string to make it `@tas` compatible
 */
export function strToSym(str: string): string {
  const ascii = anyAscii(str);
  return ascii.toLowerCase().replaceAll(/[^a-zA-Z0-9-]/g, '-');
}

export function channelHref(flag: string, ch: string) {
  return `/groups/${flag}/channels/chat/${ch}`;
}

export function makePrettyDay(date: Date) {
  const diff = differenceInDays(new Date(), date);
  switch (diff) {
    case 0:
      return 'Today';
    case 1:
      return 'Yesterday';
    default:
      return `${format(date, 'LLLL')} ${format(date, 'do')}`;
  }
}

export function makePrettyDayAndTime(date: Date) {
  const diff = differenceInDays(new Date(), date);
  const time = format(date, 'HH:mm');
  switch (true) {
    case diff === 0:
      return `Today • ${time}`;
    case diff === 1:
      return `Yesterday • ${time}`;
    case diff > 1 && diff < 8:
      return `${format(date, 'cccc')} • ${time}`;
    default:
      return `${format(date, 'LLLL')} ${format(date, 'do')} • ${time}`;
  }
}

export function makePrettyDayAndDateAndTime(date: Date) {
  const diff = differenceInDays(new Date(), date);
  const time = format(date, 'HH:mm');
  const fullDate = `${format(date, 'LLLL')} ${format(date, 'do')}, ${format(
    date,
    'yyyy'
  )}`;
  switch (true) {
    case diff === 0:
      return `Today • ${time} • ${fullDate}`;
    case diff === 1:
      return `Yesterday • ${time} • ${fullDate}`;
    case diff > 1 && diff < 8:
      return `${format(date, 'cccc')} • ${time} • ${fullDate}`;
    default:
      return `${fullDate} • ${time}`;
  }
}

export function whomIsDm(whom: ChatWhom): boolean {
  return !whom.match('/');
}
