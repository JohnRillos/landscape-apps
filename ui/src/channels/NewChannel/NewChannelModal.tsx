import React from 'react';
import Dialog, { DialogContent } from '@/components/Dialog';
import { useDismissNavigate } from '@/logic/routing';
import EditChannelForm from '../EditChannelForm';

export default function NewChannelModal() {
  const dismiss = useDismissNavigate();

  const onOpenChange = (isOpen: boolean) => {
    if (!isOpen) {
      dismiss();
    }
  };

  return (
    <Dialog defaultOpen onOpenChange={onOpenChange}>
      <DialogContent containerClass="w-full sm:max-w-lg">
        <EditChannelForm />
      </DialogContent>
    </Dialog>
  );
}
