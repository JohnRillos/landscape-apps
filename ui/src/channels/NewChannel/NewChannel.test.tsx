/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { describe, expect, it } from 'vitest';
import { render } from '../../../test/utils';
import NewChannelModal from './NewChannelModal';

vi.mock('@/state/groups', () => ({
  useRouteGroup: () => ({}),
  useAddChannelMutation: () => ({}),
}));

describe('NewChannelModal', () => {
  it('renders as expected', () => {
    const { asFragment } = render(<NewChannelModal />);
    expect(asFragment()).toMatchSnapshot();
  });
});
