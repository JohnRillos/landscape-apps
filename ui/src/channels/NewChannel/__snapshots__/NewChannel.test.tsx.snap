// Vitest Snapshot v1

exports[`NewChannel > renders as expected 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="fixed inset-0 z-30 transform-gpu bg-black opacity-30"
    data-aria-hidden="true"
    data-state="open"
    style="pointer-events: auto;"
  />
  <section
    aria-describedby="radix-2"
    aria-labelledby="radix-1"
    class="dialog-container w-full sm:max-w-lg"
    data-state="open"
    id="radix-0"
    role="dialog"
    style="pointer-events: auto;"
    tabindex="-1"
  >
    <div
      class="dialog"
    >
      <div
        class="sm:w-96"
      >
        <header
          class="mb-3 flex items-center"
        >
          <h2
            class="text-lg font-bold"
          >
            Edit Chat Channel
          </h2>
        </header>
      </div>
      <form
        class="flex flex-col"
      >
        <label
          class="mb-3 font-semibold"
        >
          Channel Name
          <input
            class="input my-2 block w-full p-1"
            name="meta.title"
            type="text"
          />
        </label>
        <label
          class="mb-3 font-semibold"
        >
          Channel Permissions
          <ul
            class="flex flex-col space-y-2"
          >
            <li>
              <label
                class="flex cursor-pointer items-center justify-between space-x-2 py-2"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="h-4 w-4 rounded-xl border-4 border-gray-400"
                  />
                </div>
                <div
                  class="flex w-full flex-col"
                >
                  <div
                    class="flex flex-row items-center space-x-2"
                  >
                    <div
                      class="flex w-full flex-col justify-start text-left"
                    >
                      <span
                        class="font-semibold"
                      >
                        Open To All
                      </span>
                      <span
                        class="text-sm font-medium text-gray-600"
                      >
                        Anyone can view and write
                      </span>
                    </div>
                  </div>
                </div>
                <input
                  class="sr-only"
                  name="privacy"
                  type="radio"
                  value="public"
                />
              </label>
            </li>
            <li>
              <label
                class="flex cursor-pointer items-center justify-between space-x-2 py-2"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="h-4 w-4 rounded-xl border-2 border-gray-200"
                  />
                </div>
                <div
                  class="flex w-full flex-col"
                >
                  <div
                    class="flex flex-row items-center space-x-2"
                  >
                    <div
                      class="flex w-full flex-col justify-start text-left"
                    >
                      <span
                        class="font-semibold"
                      >
                        Members Can View
                      </span>
                      <span
                        class="text-sm font-medium text-gray-600"
                      >
                        Members can only view
                      </span>
                    </div>
                  </div>
                </div>
                <input
                  class="sr-only"
                  name="privacy"
                  type="radio"
                  value="read-only"
                />
              </label>
            </li>
            <li>
              <label
                class="flex cursor-pointer items-center justify-between space-x-2 py-2"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="h-4 w-4 rounded-xl border-2 border-gray-200"
                  />
                </div>
                <div
                  class="flex w-full flex-col"
                >
                  <div
                    class="flex flex-row items-center space-x-2"
                  >
                    <div
                      class="flex w-full flex-col justify-start text-left"
                    >
                      <span
                        class="font-semibold"
                      >
                        Admin Only
                      </span>
                      <span
                        class="text-sm font-medium text-gray-600"
                      >
                        Only Admin can view and write
                      </span>
                    </div>
                  </div>
                </div>
                <input
                  class="sr-only"
                  name="privacy"
                  type="radio"
                  value="secret"
                />
              </label>
            </li>
          </ul>
        </label>
        <label
          class="my-2 flex justify-between font-semibold"
        >
          <div
            class="flex flex-col"
          >
            <div
              class="font-semibold"
            >
              Default
            </div>
            <div
              class="font-normal text-gray-600"
            >
              Members will join this channel automatically when joining this group
            </div>
          </div>
          <div
            class="ml-4"
          >
            <button
              aria-checked="false"
              class="switch"
              data-state="unchecked"
              role="switch"
              type="button"
              value="false"
            >
              <span
                class="switch-thumb"
                data-state="unchecked"
              />
            </button>
            <input
              aria-hidden="true"
              name="join"
              style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px;"
              tabindex="-1"
              type="checkbox"
              value="on"
            />
          </div>
        </label>
        <footer
          class="mt-4 flex items-center justify-between space-x-2"
        >
          <div
            class="ml-auto flex items-center space-x-2"
          >
            <button
              class="secondary-button ml-auto"
              type="button"
            >
              Cancel
            </button>
            <button
              class="button"
              disabled=""
              type="submit"
            >
              Done
            </button>
          </div>
        </footer>
      </form>
      <button
        class="icon-button absolute top-6 right-6"
        type="button"
      >
        <svg
          class="h-4 w-4"
          fill="none"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="stroke-current"
            d="m8 8 4-4M8 8 4 4m4 4-4 4m4-4 4 4"
            stroke-linecap="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </section>
</DocumentFragment>
`;
